drivers: dmic: Digital Microphone driver API

sathishkuttan opened the pull request:
This PR is to add an API header file for digital microphone drivers.
The APIs are defined for SoC with one or more pulse digital modulation controllers that are connected to multiple digital microphones operating as a microphone array in voice user interface applications

Commit:
drivers: dmic: APIs for digital microphones

API definition for Digital Microphone Controller(s).
Intended for SoCs with Pulse Digital Modulation controllers
configured for microphone array applications

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

rgundi commented:
@sathishkuttan .. please fix the errors reported by shippable.

hongshui3000 commented on include/dmic.h:
```diff
@@ -0,0 +1,253 @@
+ *		Sathish Kuttan <sathish.k.kuttan@.intel.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
```
SPDX-License-Identifier: BSD-3-Clause  ?

sathishkuttan replied:
Yes, the DMIC driver and APIs are being derived from sound open firmware
`https://github.com/thesofproject/sof/blob/master/src/include/sof/dmic.h`

sathishkuttan commented:
@rgundi 
The Shippable failure seems to be unrelated.
`CMake Error at /home/buildslave/.cache/zephyr/ToolchainCapabilityDatabase.cmake:166:
  Parse error.  Function missing ending ")".  End of file reached.`
Is there a way to run the sanity check locally to reproduce this?

pizi-nordic commented on include/dmic.h:
```diff
@@ -0,0 +1,253 @@
+enum dmic_trigger {
+	DMIC_TRIGGER_STOP,	/* stop stream */
+	DMIC_TRIGGER_START,	/* start stream */
+	DMIC_TRIGGER_PAUSE,	/* pause the stream */
```
What is the difference between PAUSE and STOP?
What is the difference between PAUSED and READY state?

sathishkuttan replied:
CC co-author @singalsu

`PAUSE` is intended to keep the interface running (e.g., for synchronization purposes) but pause any notifications to the applications (e.g., when the load is high and application wants to defer the processing). `RELEASE` is intended to resume the notifications.

The states are
`UNINITIALIZED` --initialize--> `READY`
`READY` --configure-->`CONFIGURED`
`CONFIGURED` --trigger start--> `ACTIVE`
`ACTIVE` --trigger pause--> `PAUSED`
`PAUSED` --trigger resume-->`ACTIVE`
`ACTIVE` --trigger stop--> `CONFIGURED`
`PAUSED` --trigger stop--> `CONFIGURED`

Perhaps, `ACTIVE` should be changed to `INITIALIZED`. Would that be better?

pizi-nordic replied:
IMHO `INITIALIZED` is better name.

pizi-nordic commented on include/dmic.h:
```diff
@@ -0,0 +1,253 @@
+ * NOTE: Only Ping-Pong buffers are supported at this time
+ */
+struct stream_buf {
+	void *buffer[2];	/* pointers to the ping pong buffers */
```
IMHO we should have buffer request/release approach, as basic ping-ping implies data copying from the statically allocated buffers.

sathishkuttan replied:
I'll add the definition for buffer request/release approach.

sathishkuttan replied:
Added `read` APIs.
Added SLAB in `configure` for the driver to allocate receive buffers from.

pizi-nordic commented on include/dmic.h:
```diff
@@ -0,0 +1,253 @@
+/**
+ * Configuration of the PCM streams to be output by the PDM hardware
+ */
+struct pcm_stream_cfg {
```
A & B are independent streams or should treat we them as a left and right channel?

sathishkuttan replied:
> A & B are independent streams or should treat we them as a left and right channel?

They are independent streams derived from the same mic. For example, from the same mic input, stream A can be generated with 48KHz/32b by using a certain decimation factor and at the same time stream B can be generated with 16KHz/16b by using a different decimation factor. This is useful for providing upto 2 streams of different **quality** from the **same microphone** to suit the application needs in terms of quality of audio.

pizi-nordic replied:
Could we be more general and allow N streams from single microphone (where N depends on the underlying HW and configuration)?

sathishkuttan replied:
@pizi-nordic Sure. Where would you define `N`?

pizi-nordic commented on include/dmic.h:
```diff
@@ -0,0 +1,253 @@
+/**
+ * Mapping/ordering of the PDM channels to logical PCM output channel
+ */
+struct pdm_chan_cfg {
```
What about controllers which have fixed channel mapping?
Also, since there are multiple controllers could we handle them independently?

sathishkuttan replied:
> What about controllers which have fixed channel mapping?

I could add a channel map output parameter to return the **actual** mapping that the driver is able to create on the underlying hardware.

> Also, since there are multiple controllers could we handle them independently?

Theoretically, they could be independent if the hardware supports it (the hardware I use does not). The definition is geared towards **array** applications, so they are intended to be used as a collection of mics that are synchronized. The driver or the application on top of the driver could still separate the streams to make it look like they are independent streams.

pizi-nordic commented:
CC: @pdunaj, @carlescufi

pdunaj commented:
Can you quickly tell how application is notified that data was recorded?

sathishkuttan commented:
> Can you quickly tell how application is notified that data was recorded?

via the callbacks `stream_a_callback` and `stream_b_callback` that are provided by the application in `struct pcm_stream_cfg`

sathishkuttan commented:
Removed callbacks and added a `read` API.
Changed the API to allow applications to request arbitrary number of streams. Driver will return the actual number of streams that it could enable based on underlying hardware.

pdunaj commented on include/audio/dmic.h:
```diff
@@ -0,0 +1,301 @@
+struct _dmic_ops {
+	int (*configure)(struct device *dev, struct dmic_cfg *config);
+	int (*trigger)(struct device *dev, enum dmic_trigger cmd);
+	int (*read)(struct device *dev, u8_t stream, void **buffer,
```
Read is fine but I have one comment.
In typical case application would have to create a thread to wait for data. If there is more then one stream, there will have to be one dedicated thread for one stream. This is problematic as each thread has to have its own stack causing memory waste.

I don't think there is a generic API for poll in Zephyr but maybe we can start here. Can we add a function to wait for any stream to be ready (*wait_for_streams(..*dev, u8_t *streams, size_t *count)), that would return when data is available in any stream. Arguments: stream - array to hold ids of streams ready, count - number entries in array when called, number of stream ready when returned back. This way thread would call it, and only upon its return would call read on the streams that are ready.

@carlescufi , @pizi-nordic This look like another example when poll similar to linux would be useful.

sathishkuttan replied:
> If there is more then one stream, there will have to be one dedicated thread for one stream. This is problematic as each thread has to have its own stack causing memory waste.

It's not problematic. The streams are application specific (voice calling application will require lower sample rate compared to a high quality audio recording for reproduction purposes). So having a read with stream id that the respective application can read using it's own thread is going to be use case.

pdunaj replied:
@sathishkuttan I don't agree. Everything depends on the application. It may be that one would like to read two streams on SoC with low RAM and (with this API) will be forced to waste memory for additional thread's stack.
If polling (in whatever form) is present, application can use it (to read many streams from single thread) or don't (create a thread or even an application for one stream). But it is an application choice.
Anyway this is a general problem in Zephyr.

sathishkuttan replied:
@pdunaj You don't have to agree. You are entitled to your opinion.

pdunaj replied:
@sathishkuttan Thanks ;)

sathishkuttan replied:
@pdunaj You are welcome :)
By the way, this API does not force to waste memory. In my actual implementation, I am reading 2 streams from one thread -- one from this DMIC and another from I2S. The thread blocks on DMIC read and once the data is read from DMIC, the I2S data is read without a wait. Through a proper hardware scheduling, I ensure that the I2S data is ready when the DMIC data is ready.
While in this case, it is between DMIC and I2S, the same concept can be used for 2 DMIC streams.

pdunaj commented on include/audio/dmic.h:
```diff
@@ -0,0 +1,301 @@
+};
+
+/**
+ * @brief Build the channel map to populate struct pdm_chan_cfg
```
Doxygen brief explicitly used while previous elements had brief done using javadoc autobrief style.
I don't mind any but its better to be consistent.

sathishkuttan replied:
Thanks for the input. I'll make this consistent.

pizi-nordic commented on include/audio/dmic.h:
```diff
@@ -0,0 +1,301 @@
+	/* parameters global to all PDM controllers */
+
+	/* minimum clock frequency supported by the mic */
+	u32_t	min_pdm_clk_freq;
```
How clock frequency and duty cycle is determined? Could we just have per-mic/per-controller configuration?

sathishkuttan replied:
They are specified in the datasheet of the microphones.
We could have per-mic, but it is unnecessary. Typically all microphones in an array are from the same vendor and same type. It's an overkill to define the API to allow this to be specified for each mic individually especially if the mic count is high.

pizi-nordic commented on include/audio/dmic.h:
```diff
@@ -0,0 +1,301 @@
+	u32_t			pcm_rate;
+	/* PCM sample width of stream */
+	u8_t			pcm_width;
+	u8_t			reserved;
```
The reserved field is not needed.

sathishkuttan replied:
It was added for alignment. If convention is to not add such fields, I can remove it.

pizi-nordic commented on include/audio/dmic.h:
```diff
@@ -0,0 +1,301 @@
+ *
+ * @return Bit-map containing the PDM and L/R channel information
+ */
+static inline u32_t dmic_build_channel_map(u8_t channel, u8_t pdm,
```
The encoding in functions below seems to be HW-specific. In my opinion at this level we have to use a generic approach.

sathishkuttan replied:
It is generic. PDM mics have a left/right control signal. Please refer to publicly available datasheets of PDM mics from different vendors.

pizi-nordic commented on include/audio/dmic.h:
```diff
@@ -0,0 +1,301 @@
+ *
+ * @return 0 on success, a negative error code on failure
+ */
+static inline int dmic_trigger(struct device *dev, enum dmic_trigger cmd)
```
IMHO dmic_cmd() is a better name. 
Also, there is no callback informing when given command is executed (I assume that some controllers might delay stop command until a memory buffer will be fully filled).

sathishkuttan replied:
This follows the convention used in existing I2S driver, DMA driver and potentially other drivers. I prefer to be consistent with other drivers.

pizi-nordic reviewed (commented):
The stream, controller and channel mapping is still unclear for me. Could we use a more generic approach at this layer?

Something like this:
```
Stream #0: Controller #0, Channel #2, Left
Stream #1: Controller #0, Channel #2, Left
Stream #2: Controller #1, Channel #1, Stereo
```
Also, I do not see any information about stream format (data layout in case of stereo stream).

nashif reviewed (changes requested):
this would require license approval, can't be merged as is.

sathishkuttan commented:
> this would require license approval, can't be merged as is.

changed license to Apache 2.0
